lib/std/os/windows.zig

```diff
@@ -952,12 +952,64 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
     return @bitCast(u64, result);
 }
 
-pub const GetFinalPathNameByHandleError = error{
-    BadPathName,
-    FileNotFound,
-    NameTooLong,
-    Unexpected,
-};
+pub fn QueryObjectName(
+    Handle: HANDLE,
+    out_buffer: []u16,
+    ReturnLength: ?PULONG, //returned in bytes
+) ![:0]u16 {
+    const alignment = @alignOf(OBJECT_NAME_INFORMATION);
+    const begin_addr = @ptrToInt(out_buffer.ptr);
+    const aligned_begin_addr = std.mem.alignForward(begin_addr, alignment);
+    const alignment_step = aligned_begin_addr - begin_addr;
+    const aligned_raw_buffer = @alignCast(alignment, std.mem.sliceAsBytes(out_buffer)[alignment_step..out_buffer.len*@sizeOf(u16)]);
+    if(aligned_raw_buffer.len < @sizeOf(OBJECT_NAME_INFORMATION)) return error.BufferTooSmall; //this would result in .INFO_LENGTH_MISMATCH
+    
+    var info = @ptrCast(*OBJECT_NAME_INFORMATION, aligned_raw_buffer);
+    info.* = undefined; //all of the fields we read will be overwritten, and the result string is placed in the same buffer after the struct
+    const out_buffer_length = std.math.cast(ULONG, aligned_raw_buffer.len) catch |e| std.math.maxInt(u32); //we'll just use as much of it as we can
+    const rc = ntdll.NtQueryObject(Handle, .ObjectNameInformation, out_buffer.ptr, out_buffer_length, ReturnLength); //buffer size is specified in u16...
```

Comment:
The code is extremely hairy, it takes an output buffer `out_buffer` where the pathname is stored _and_ uses it to allocate a `OBJECT_NAME_INFORMATION` too.

I'd simply allocate a buffer that's hopefully large enough (path_max plus the `OBJECT_NAME_INFORMATION` overhead) and copy the result over.